functions-shortcodes.php

```diff
@@ -0,0 +1,253 @@
+add_filter('the_content', 'simplecommerce_restore_shortcodes_after_wpautop', 10);
+
+function simplecommerce_rescue_shortcodes_from_wpautop( $content ) {
+    global $shortcode_replacements;
+
+    function find_first_shortcode_pos( $content, $offset ) {
+        global $shortcodes;
+
+        $earliest_pos = -1;
+        $first_shortcode = null;
+
+        foreach ( $shortcodes as $shortcode ) {
+            $pos = strpos( $content, '[' . $shortcode . ']', $offset );
+
+            if ( $pos !== false && ( $earliest_pos == -1 || $pos < $earliest_pos ) ) {
+                $earliest_pos = $pos;
+                $first_shortcode = $shortcode;
+            }
+        }
+
+        return $earliest_pos;
+    }
+
+    // Search for any instance of known shortcodes.
+    // Upon finding a shortcode opening tag, continue searching forward for the corresponding closing tag.
+    $pos = 0;
+
+    while ( $pos < strlen( $content ) ) {
+        $pos = find_first_shortcode_pos( $content, $pos );
+
+        if ( $pos == -1 ) {
+            break;
+        }
+
+        // Found an opening tag? Continue searching until we find the matching closing tag.
+        // There is the possibility of nested opening tags before reaching the matching close.
+        $tag_name = substr( $content, $pos + 1, strpos( $content, ']', $pos ) - $pos - 1 );
+        $opening_tag_regex = '/\[' . $tag_name . '( [^]]+)?\]/';
+        $closing_tag = '[/' . $tag_name . ']';
+        $open_tag_count = 1;
+        $current_index = $pos;
+
+        while ( $open_tag_count > 0 && $current_index > -1 && $current_index < strlen( $content )) {
+            $closing_pos = strpos( $content, $closing_tag, $current_index );
+            $open_tag_found = preg_match( $opening_tag_regex, $content, $matches, PREG_OFFSET_CAPTURE, $current_index );
+
+            if ( $open_tag_found ) {
+                $matched_open_tag = $matches[0][0];
+                $next_opening_pos = $matches[0][1];
+            }
+
+            if ( $closing_pos === false ) {
+                $closing_pos = -1;
+            }
+
+            if ( !$open_tag_found || $closing_pos < $next_opening_pos ) {
+                if ( $closing_pos == -1 ) {
+                    // This is a problem; we have no closing tag. Just take the rest of the content.
+                    $open_tag_count = 0;
+                    $closing_pos = strlen( $content ) - strlen( $closing_tag );
+                    break;
+                }
+
+                $current_index = $closing_pos + strlen( $closing_tag );
+                $open_tag_count--;
+            } else {
+                $current_index = $next_opening_pos + strlen( $matched_open_tag );
+                $open_tag_count++;
+            }
+        }
+
+        // Replace everything from $pos to $current_index with a placeholder and stash that placeholder.
+        // Don't forget to wrap it in <p> so that wpautop doesn't. :P
+        $key = '<p>' . md5(rand()) . '</p>';
```

Comment:
Add a check against existing `$shortcode_replacements` to ensure `$key` is unique.